src/NuGet.Core/NuGet.Protocol/Model/GetVulnerabilityInfoResult.cs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol.Model
+{
+    /// <summary>Represents the result of <see cref="IVulnerabilityInfoResource.GetVulnerabilityInfoAsync(Core.Types.SourceCacheContext, Common.ILogger, System.Threading.CancellationToken)"/></summary>
+    /// <remarks>Vulnerability data from a package source typically is spread across multiple files. Therefore, it's
+    /// possible for some data to be retrieved successfully, while other data failed. This type allows partial data
+    /// to be returned along with error information.</remarks>
+    public class GetVulnerabilityInfoResult
+    {
+        /// <summary></summary>
+        public IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? KnownVulnerabilities { get; }
+        public IReadOnlyList<Exception>? Exceptions { get; }
+
+        public GetVulnerabilityInfoResult(
```

Comment:
Since GitHub doesn't provide threaded/resolvable conversations unless assigned to a specific line of code:

From @erdembayar in https://github.com/NuGet/NuGet.Client/pull/5115#issuecomment-1492755715

> If performance is a concern here then have you considered other data structure types? Perhaps a Trie data structure could be an alternative instead of a big nested loop. We have already implemented one for [Package Source mapping lookup](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/SearchNode.cs) by pattern. Actually, it could be further improved by using string compression. I haven't had time to implement this, and so far there have been no reported performance concerns regarding that.
The only high cost would be the initial Trie tree building, and then the lookup would be cheap. Still tbh, without actual perf test it's hard to say what would be perf difference, it depends on parameters like load size and number of packages etc.

From @kartheekp-ms in https://github.com/NuGet/NuGet.Client/pull/5115#issuecomment-1492769962

> In my opinion, the Trie data structure may not be appropriate for this scenario as our objective is not to match prefixes, as in the case of a package source mapping feature.

As I wrote in [the PR review suggestions](https://github.com/NuGet/NuGet.Client/pull/5115#issuecomment-1488678765), under a header titled "Why not merge multiple files into a single dictionary"

> benchmarking makes it appear that the time to merge the multiple dictionaries into a single dictionary, or reading the vulnerabilities files serially into a single dictionary, rather than in parallel into multiple dictionaries, is slower than having the vulnerability checking code iterate over multiple dictionaries.

Putting it another way, the perf-bottleneck is merge the data into a single data structure, not making many lookups for a single package ID.

Reply:
> Putting it another way, the perf-bottleneck is merge the data into a single data structure, not making many lookups for a single package ID.

I agree with that. Actually, building the initial tree for trie is expensive, but lookups would be quite cheap. The reason I propose using trie is that, if I'm not mistaken, for a large set of package IDs it would save space since the size does not grow linearly with the number of items in it. Therefore, it solves the problem of vulnerability data growing over time.

If the internal implementation is abstracted away and changing it wouldn't cause a public API change later, then I'm okay with the current implementation.

Reply:
It is something that can change later, but I feel like it's missing the point. Synthetic benchmarks so far suggest that loading the file takes more time than lookups. So, if we make lookups even faster, by making loading even slower, then it's going to make the total duration worse, not better. We can validate once the real implementation is complete to validate that the synthetic benchmarks were representative of real performance.